### PR TITLE
[Relations]: solve swapping entity ids without unmounting components

### LINKED ISSUE #15638: Relation fields are broken in non-localized dynamic zone
### Required System information

- Node.js version: 18.13.0
- NPM version: 8.19.3
- Strapi version: 4.6.0
- Database: PostgreSQL
- Operating system: Ubuntu 22.04

### Describe the bug

I have a localized content type (article) with a dynamic zone (restrictions) which should be common for all locales. 
image

When I add a component in the dynamic zone with a relation, save the article, then switch to another locale, the relation is empty:
image

This only affects relation fields, other fields (text in my example) are prefilled correctly. Now, if I try to edit the relation for the locale the Strapi UI crashes completely:

image

The only workaround is to localize the dynamic zone, but this is obviously not desirable.

### LINKED ISSUE #15743: Dynamic zones with relations inside components do not update correctly when switching locales and causes blank page
### Required System information

- Node.js version: 16.14.0
- NPM version: 8.3.1
- Strapi version: 4.6.0
- Database: mysql
- Operating system: CentOS / Local MacOS
- Is your project Javascript or Typescript: Javascript

### Describe the bug

When switching between locales (for example Dutch > English) the dynamic zone components of the Dutch locale stay visible on the English locale, eventhough the English locale does not have this component added in that particular dynamic zone. If you then refresh the page, the selected locale is correctly displayed (in this case without the component that is used on the Dutch locale). But then when I switch back to the Dutch locale, the page turns blank and displays the following error in the console:

`TypeError: Cannot read properties of null (reading 'length')`.

The error is inside `RelationInputDataManager.js` on line 58 -> `const currentLastPage = Math.ceil(get(initialData, name, []).length / RELATIONS_TO_DISPLAY);`

If I remove the relation from the component this issue is resolved. 

### Steps to reproduce the behavior

See video's that I added to this post

### Expected behavior

When you switch locales the dynamic zones should be updated according to what's filled in for that particular locale. 

### Screenshots

See video's that i added to this post.

### Code snippets

n.a.

### Additional context

There's a difference in errors whether the relation is filled or not. I added two video's outlining the issues

https://user-images.githubusercontent.com/43511726/217519311-4c041155-fd40-42ed-b28d-b38c26ebd20a.mov

https://user-images.githubusercontent.com/43511726/217519350-c7cbbab7-df98-4afb-8934-8635a6fcd7f7.mov

### PR DESCRIPTION
**Pull request by joshuaellis**
### What does it do?

* Merge the existing data with new data only if the entity ids are the same

### Why is it needed?

* relations are cached between locales which is typically undesired.

### How to test it?

* Create a localised content type
* Add a relation field
* Set the relation in one locale & save
* refresh the page
* see this relation correctly there
* swap locale
* the relation field should be empty

There's also an automated test.

### Related issue(s)/PR(s)

* resolves #15743
* resolves #15638
* resolves CONTENT-1032

### Thoughts

* ~Is there a way to pass this string to both places so we dont rely on spelling? 🤔  cc @gu-stav~
* ~Who's responsibility is this? It feels over eager for the content-manager to dismiss the cache when any part of the URL changes? But then if someone else creates a plugin _similar_ to how i18n works then they'll need to add this code too...~

### CODE REVIEW

**Review by gu-stav — COMMENTED** [review]
Just an idea: wouldn't it be way easier to include `defaultParams` as the query cache-key? `defaultParams` will contain all injected plugin params and if we'd include it in the key, we'd re-fetch every time when a locale was changed I think.

**Comment by joshuaellis** [comment]
> Just an idea: wouldn't it be way easier to include `defaultParams` as the query cache-key? `defaultParams` will contain all injected plugin params and if we'd include it in the key, we'd re-fetch every time when a locale was changed I think.

So i guess to ensure the plugin contract is met we have to pass _all_ `defaultParams`. My concern would be overfetching... but maybe that's something we have to try and see?

**Comment by gu-stav** [comment]
Yeah, there is a slight chance of over-fetching, but given how much easier this approach seems to me at first it might be worth giving it a shot. I haven't seen any other plugin so what which makes use of this not-very-well documented capability.

File: packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/reducer.js [hunk 1]
```diff
@@ -197,6 +197,14 @@ const reducer = (state, action) =>
...
         break;
       }
+      case 'CLEAR_RELATION': {
+        const { initialDataPath, modifiedDataPath } = action;
+
+        set(draftState, initialDataPath, []);
+        set(draftState, modifiedDataPath, []);
+
+        break;
+      }
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/reducer.js:207]: I know it doesn't add much value at the moment, but could you add a test for this case? 🙏🏼

File: packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js [hunk 2]
```diff
@@ -125,6 +125,19 @@ export const useRelation = (cacheKey, { relation, search }) => {
...
   const onLoadRelationsCallback = useCallbackRef(onLoadRelations);
 
+  const clearRelationsCallback = useCallbackRef(clearRelations);
+
+  /**
+   * This fires on unmount only.
+   */
+  useEffect(() => {
+    return () => {
+      if (clearRelationsCallback) {
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js:135]: The idea looks good to me - good thinking. What do you think about renaming the callback to `onUnmount`, though?

**Review by gu-stav — COMMENTED** [review]
Unfortunately there seems to be a bug when using a relation inside of a repeatable component:

https://user-images.githubusercontent.com/2244375/221888403-353e21bb-1423-4eb5-b6aa-b566736fa606.mp4

**Comment by joshuaellis** [comment]
I'd love to solve a way of automating these tests, but I can't think of any.... 😔

**Comment by joshuaellis** [comment]
Hey @Marc-Roig would you be able to QA this ticket please? I changed my approach which imo minimised potential impact so I would do the following:

* Add relations in DynamicZones / Repeatable Components & at the root level of the entry
* Test with swapping locales for your entry

The user story overall would be "As a user I should be able to swap locales on entities with relations and not have the relation fields pollute each other"

Then @jhoward1994 if you could focus primarily on a code review that would be great 🙏🏼

File: examples/getstarted/src/api/relation-locale/content-types/relation-locale/schema.json [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      "type": "component",
+      "repeatable": true,
+      "pluginOptions": {
+        "i18n": {
+          "localized": true
+        }
+      },
+      "component": "basic.relation"
+    },
+    "dyanmic_relations": {
```
Comment (jhoward1994) [inline examples/getstarted/src/api/relation-locale/content-types/relation-locale/schema.json:43]:
```suggestion
    "dynamic_relations": {
```

**Comment by jhoward1994** [comment]
The code looks good to me, happy to approve this if the QA is successful 👍🏻

**Pull request merged**